py/terra_widgets/html_snapshots.py

```diff
@@ -289,33 +291,37 @@ def on_choose_workspace(changed):
     with output:
       output.clear_output()
       workspace_paths = ws_paths[changed['new']]
-      comment_files = get_ipython().getoutput(f'gsutil ls {workspace_paths.get_comment_file_glob()}')
-      if not comment_files[0].startswith('gs://'):
+      try:
+        comment_files = tf.io.gfile.glob(pattern=workspace_paths.get_comment_file_glob())
+      except tf.errors.PermissionDeniedError as e:
+        target_workspace = [name for name, id in ws_names2id.items() if id == changed['new']]
+        display(HTML(f'''<div class="alert alert-block alert-danger">
+          <b>Warning:</b> Unable to view HTML snapshots in workspace {target_workspace} from <b>this workspace</b>.
+          <hr><p><pre>{e.message}</pre></p>
+          </div>'''))
+        return
+      if not comment_files:
         display(HTML('''<div class="alert alert-block alert-warning">
           No comment files found for HTML snapshots in this workspace.</div>'''))
         return
-      progress = widgets.IntProgress(
-          value=0,
-          min=0,
-          max=len(comment_files),
-          step=1,
-          description=f'Retrieving {len(comment_files)} comments:',
-          bar_style='success',
-          orientation='horizontal',
-          layout=widgets.Layout(width='450px'),
-          style={'description_width': 'initial'}
-      )
-      display(progress)
-      comment_num = 0
-      comment_file_contents = []
-      for file in comment_files:
-        comment = get_ipython().getoutput(f"gsutil cat '{file}'")
-        version = file.replace(workspace_paths.get_subfolder(), '')
-        comment_file_contents.append({'version': version, 'comment': comment})
-        comment_num += 1
-        progress.value = comment_num
-      comments = pd.DataFrame(comment_file_contents)
-      display(comments)
+
+      def get_comment(f):
+        with tf.io.gfile.GFile(f, 'r') as fh:
+          return fh.readlines()
+
+      def process_task(f):
+        return f, get_comment(f)
+
+      max_pool = 8
+      with Pool(max_pool) as p:
+        pool_outputs = list(tqdm(p.imap(process_task, comment_files), total=len(comment_files)))
```

Comment:
Checking my understanding here: `imap` vs `map` - `imap` documentation is poor, but my guess is this yields elements as they complete, which thereby allows you to leverage the tqdm progress bar? (vs map which I imagine blocks for everything)

Reply:
Yes, that matches my understanding. Thanks for the review!